app/gui2/ydoc-server/languageServerSession.ts

```diff
@@ -466,13 +471,21 @@ class ModulePersistence extends ObservableV2<{ removed: () => void }> {
     if (this.syncedContent == null || this.syncedVersion == null) return
 
     const code = newCode ?? synced.code
-    const newMetadataJson =
-      newMetadata &&
-      json.stringify({ ...this.syncedMeta, ide: { ...this.syncedMeta.ide, node: newMetadata } })
-    const newIdMapJson = newIdMap && serializeIdMap(newIdMap)
+    const metadataToPersist = {
+      ...this.syncedMeta,
+      ide: { ...this.syncedMeta.ide, node: newMetadata },
+    }
+    const newMetadataJson = newMetadata && json.stringify(metadataToPersist)
+    const idMapToPersist =
+      newIdMap &&
```

Comment:
If the metadata line changes the IdMap will be recomputed as well. And if the metadata was changed without changing the IDs, we don't need to edit the IdMap section.

Reply:
> If the metadata line changes, the IdMap will be recomputed as well.

I don't see this being enforced in the ydoc-server. If you mean it will be recomputed at the same time in the GUI, the ydoc-server should not depend on that.

When adding a new node, first the new node is committed then its position is computed. In the interim, the node usually doesn't have any non-default metadata; if we are always writing node metadata at the same time that a node is added, that's an inefficiency we should be able to fix in the GUI without the ydoc-server causing data loss.

Reply:
I can see that `applyDocumentUpdates`
https://github.com/enso-org/enso/blob/305f69cd49ec8dfcedec0c3dfe9f8247a13da80c/app/gui2/ydoc-server/languageServerSession.ts#L451-L455
recomputes IdMap from the `root` AST
https://github.com/enso-org/enso/blob/305f69cd49ec8dfcedec0c3dfe9f8247a13da80c/app/gui2/ydoc-server/edits.ts#L67-L74
or am I reading that wrong?

Reply:
The condition `codeChanged || idsChanged || synced.idMapJson == null` can still be false when existing metadata structure updates. In that situation a new ID will have to be stored while no code was actually changed.

Currently, all metadata is written on expressions that also happen to be node roots, so it kinda works out. But this is not a limitation of the system that we are willing to have, since we are soon going to add metadata to individual expressions. That will be needed to store persistent widget state. Those expressions will already exist when we adjust the widget data, effectively sending a delaying metadata update without any code changes.

Reply:
Updated